src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,199 @@
+        Node<T> nodeAtIndex = getNodeAtIndex(index);
+
+        if (nodeAtIndex != null) {
+            T oldValue = nodeAtIndex.value;
+            nodeAtIndex.value = value;
+            return oldValue;
+        }
+
         return null;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexValid(index);
+        Node<T> nodeToRemove = getNodeAtIndex(index);
+        T valueToRemove = nodeToRemove.value;
+        removeNode(nodeToRemove);
+        size--;
+        return valueToRemove;
     }
 
-    @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+
+        while (current != null) {
+            if (Objects.equals(object, current.value)) {
+                removeNode(current);
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
+    private void addFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        if (head != null) {
+            head.prev = newNode;
+        }
+        head = newNode;
+        if (tail == null) {
+            tail = newNode;
+        }
+        size++;
+    }
+
+    // Helper method to add an element at the end of the list
+    private void addLast(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (head == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
+    }
+
+    private void removeNode(Node<T> node) {
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        } else {
+            tail = node.prev;
+        }
+    }
+
+    private Node<T> getNodeAtIndex(int index) {
+        isIndexValid(index);
+        Node<T> current;
+        int currentIndex;
+
+        if (index > size() / 2) {
+            current = tail;
+            currentIndex = size() - 1;
+            while (currentIndex > index) {
+                current = current.prev;
+                currentIndex--;
+            }
+        } else {
+            current = head;
+            currentIndex = 0;
+            while (currentIndex < index) {
+                current = current.next;
+                currentIndex++;
+            }
+        }
+        return current;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private void isIndexOutOfArray(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Index is out of range: "
+                    + index + " for size: " + size());
+        }
+    }
+
+    private void isIndexValid(int index) {
```

Comment:
Why did you duplicate this method? You've already created `isIndexOutOfArray` but there is no difference in their behavior.

Reply:
Once im calling that index is out of bounds, second one calls for invalid index, forgot how to implement it when its called(to change messege when i call it).

Reply:
i change it, so i call only one of those methods without changing message